File: freeq-sdk/src/e2ee_group.rs
//! Sender-keys group encryption with VC-gated, server-blind key distribution (EG1).
//!
//! This is the intended replacement for the broken [`crate::e2ee_did`] `GroupKey`
//! (ENC2), whose key was derived purely from *public* data (member DIDs + channel
//! name), so the server — or anyone who can enumerate membership — could
//! recompute it. That scheme provides **no** confidentiality against the host.
//!
//! EG1 fixes this: the channel's group key is a **random 32-byte secret** that
//! only members ever hold. It is distributed to each member out-of-band of the
//! server's view by *sealing* it to that member's X25519 public key (published
//! in their pre-key bundle). The server relays the sealed blob but can never
//! open it. Membership is gated by the policy/VC framework — the "key steward"
//! only seals to a member after their JOIN cleared policy (e.g. a Google/SAML
//! credential), so authorization and key access share one source of truth.
//!
//! # Roles
//!
//! - **Steward** — an op client (or a company-run bot) that holds the current
//!   [`GroupState`] and seals it to each admitted member. There is no server
//!   custody of the key.
//! - **Member** — holds a long-lived X25519 identity key. Receives a
//!   [`SealedGroupKey`] control message, [`GroupState::open`]s it, then
//!   encrypts/decrypts channel traffic with the recovered [`GroupState`].
//!
//! # Wire formats
//!
//! Channel message (rides in a `+encrypted` PRIVMSG, like ENC1/ENC2):
//! ```text
//! EG1:<epoch>:<nonce-b64>:<ciphertext-b64>
//! ```
//!
//! Sealed key-wrap control message (steward → one member):
//! ```text
//! EGK1:<channel>:<epoch>:<ephemeral-pub-b64>:<nonce-b64>:<ciphertext-b64>
//! ```
//!
//! # Rotation & revocation (the property ENC1/ENC2 lack)
//!
//! On any membership change (leave, kick, or VC expiry — e.g. an offboarded
//! employee), the steward calls [`GroupState::rotate`] to mint a *fresh* random
//! secret at `epoch+1` and re-seals only to the remaining members. The departed
//! member never receives the new epoch's key, so they cannot read new traffic —
//! forward secrecy against membership change. Old ciphertext they already saw
//! stays readable to them (that is inherent to any group scheme without
//! per-message ratcheting; see the doc for the MLS upgrade path).

use aes_gcm::aead::{Aead, KeyInit, OsRng};
use aes_gcm::{AeadCore, Aes256Gcm, Nonce};
use base64::Engine;
use base64::engine::general_purpose::URL_SAFE_NO_PAD as B64;
use hkdf::Hkdf;
use rand::RngCore;
use sha2::{Digest, Sha256};
use x25519_dalek::{PublicKey, StaticSecret};

/// Channel-message prefix (EG1 = "encrypted group v1").
pub const EG1_PREFIX: &str = "EG1:";
/// Sealed key-wrap control-message prefix.
pub const EGK1_PREFIX: &str = "EGK1:";

/// The confidential state a member needs to read/write a channel at one epoch.
///
/// The `secret` is 32 bytes of CSPRNG output — **not** derived from any public
/// value. Cloneable so a steward can seal the same epoch to many members.
#[derive(Clone)]
pub struct GroupState {
    /// Channel name (lowercased for domain separation).
    pub channel: String,
    /// Key epoch; increments on every membership change.
    pub epoch: u64,
    /// The random group secret. Never leaves a member's device unsealed.
    secret: [u8; 32],
}

impl GroupState {
    /// Steward: create a brand-new group at epoch 1 with a random secret.
    pub fn create(channel: &str) -> Self {
        let mut secret = [0u8; 32];
        OsRng.fill_bytes(&mut secret);
        Self {
            channel: channel.to_lowercase(),
            epoch: 1,
            secret,
        }
    }

    /// Steward: mint the next epoch with a fresh random secret. Call this on
    /// every membership change, then re-seal to the *remaining* members only.
    pub fn rotate(&self) -> Self {
        let mut secret = [0u8; 32];
        OsRng.fill_bytes(&mut secret);
        Self {
            channel: self.channel.clone(),
            epoch: self.epoch + 1,
            secret,
        }
    }

    /// Per-epoch, per-channel AES-256 message key. Domain-separates epochs so a
    /// nonce collision across epochs can't cross-decrypt, and binds ciphertext
    /// to the channel it was sent in.
    fn message_key(&self) -> [u8; 32] {
        let salt = Sha256::digest(self.channel.as_bytes());
        let hk = Hkdf::<Sha256>::new(Some(&salt), &self.secret);
        let info = format!("freeq-group-msg-v1-{}", self.epoch);
        let mut key = [0u8; 32];
        hk.expand(info.as_bytes(), &mut key)
            .expect("32 is a valid HKDF-SHA256 length");
        key
    }

    /// Encrypt a channel message → `EG1:<epoch>:<nonce>:<ct>`.
    pub fn encrypt(&self, plaintext: &str) -> Result<String, GroupError> {
        let key = self.message_key();
        let cipher = Aes256Gcm::new_from_slice(&key).map_err(|_| GroupError::BadKey)?;
        let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
        let ct = cipher
            .encrypt(&nonce, plaintext.as_bytes())
            .map_err(|_| GroupError::Crypto)?;
        Ok(format!(
            "{EG1_PREFIX}{}:{}:{}",
            self.epoch,
            B64.encode(&nonce[..]),
            B64.encode(&ct)
        ))
    }

    /// Decrypt an `EG1:` channel message. Errors on epoch mismatch so the caller
    /// knows it needs the sealed key for a different epoch.
    pub fn decrypt(&self, wire: &str) -> Result<String, GroupError> {
        let body = wire.strip_prefix(EG1_PREFIX).ok_or(GroupError::NotEg1)?;
        let parts: Vec<&str> = body.splitn(3, ':').collect();
        if parts.len() != 3 {
            return Err(GroupError::Malformed);
        }
        let epoch: u64 = parts[0].parse().map_err(|_| GroupError::Malformed)?;
        if epoch != self.epoch {
            return Err(GroupError::EpochMismatch {
                have: self.epoch,
                got: epoch,
            });
        }
        let nonce_bytes = B64.decode(parts[1]).map_err(|_| GroupError::Malformed)?;
        let ct = B64.decode(parts[2]).map_err(|_| GroupError::Malformed)?;
        if nonce_bytes.len() != 12 {
            return Err(GroupError::Malformed);
        }
        let key = self.message_key();
        let cipher = Aes256Gcm::new_from_slice(&key).map_err(|_| GroupError::BadKey)?;
        let pt = cipher
            .decrypt(Nonce::from_slice(&nonce_bytes), ct.as_ref())
            .map_err(|_| GroupError::Crypto)?;
        String::from_utf8(pt).map_err(|_| GroupError::Utf8)
    }

    /// Steward: seal this epoch's secret to one member's X25519 public key.
    ///
    /// Uses ephemeral-static ECIES: a fresh ephemeral keypair per seal, ECDH to
    /// the member's static key, HKDF to a wrapping key, AES-256-GCM over the
    /// secret. The ephemeral public key travels in the clear (it's public). Only
    /// the holder of the member's static secret can rederive the wrapping key.
    pub fn seal_for(&self, member_pub: &[u8; 32]) -> SealedGroupKey {
        let eph_secret = StaticSecret::random_from_rng(OsRng);
        let eph_pub = PublicKey::from(&eph_secret);
        let member_pub = PublicKey::from(*member_pub);
        let shared = eph_secret.diffie_hellman(&member_pub);

        let wrap_key = wrap_key(shared.as_bytes(), &self.channel, self.epoch);
        let cipher = Aes256Gcm::new_from_slice(&wrap_key).expect("32-byte key");
        let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
        let ct = cipher
            .encrypt(&nonce, self.secret.as_ref())
            .expect("AES-GCM seal cannot fail on valid inputs");

        SealedGroupKey {
            channel: self.channel.clone(),
            epoch: self.epoch,
            ephemeral_pub: eph_pub.to_bytes(),
            nonce: nonce.into(),
            ciphertext: ct,
        }
    }

    /// Member: recover the group state from a sealed key-wrap using your X25519
    /// static secret. This is the only way to obtain the secret — the server,
    /// which relays the [`SealedGroupKey`], cannot.
    pub fn open(sealed: &SealedGroupKey, my_secret: &StaticSecret) -> Result<Self, GroupError> {
        let eph_pub = PublicKey::from(sealed.ephemeral_pub);
        let shared = my_secret.diffie_hellman(&eph_pub);
        let wrap_key = wrap_key(shared.as_bytes(), &sealed.channel, sealed.epoch);
        let cipher = Aes256Gcm::new_from_slice(&wrap_key).map_err(|_| GroupError::BadKey)?;
        let pt = cipher
            .decrypt(Nonce::from_slice(&sealed.nonce), sealed.ciphertext.as_ref())
            .map_err(|_| GroupError::Crypto)?;
        let secret: [u8; 32] = pt.try_into().map_err(|_| GroupError::Malformed)?;
        Ok(Self {
            channel: sealed.channel.clone(),
            epoch: sealed.epoch,
            secret,
        })
    }
}

/// A group secret sealed to exactly one member's X25519 public key.
///
/// Safe for the server to store and relay: without the member's static secret
/// it is an opaque AEAD blob.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SealedGroupKey {
    pub channel: String,
    pub epoch: u64,
    pub ephemeral_pub: [u8; 32],
    pub nonce: [u8; 12],
    pub ciphertext: Vec<u8>,
}

impl SealedGroupKey {
    /// Serialize to `EGK1:<channel>:<epoch>:<eph-pub>:<nonce>:<ct>`.
    pub fn to_wire(&self) -> String {
        format!(
            "{EGK1_PREFIX}{}:{}:{}:{}:{}",
            self.channel,
            self.epoch,
            B64.encode(self.ephemeral_pub),
            B64.encode(self.nonce),
            B64.encode(&self.ciphertext),
        )
    }

    /// Parse an `EGK1:` control message. Channel names cannot contain ':', which
    /// holds for IRC channels, so a fixed field split is unambiguous.
    pub fn from_wire(wire: &str) -> Result<Self, GroupError> {
        let body = wire.strip_prefix(EGK1_PREFIX).ok_or(GroupError::NotEgk1)?;
        let parts: Vec<&str> = body.splitn(5, ':').collect();
        if parts.len() != 5 {
            return Err(GroupError::Malformed);
        }
        let channel = parts[0].to_string();
        let epoch: u64 = parts[1].parse().map_err(|_| GroupError::Malformed)?;
        let ephemeral_pub: [u8; 32] = B64
            .decode(parts[2])
            .map_err(|_| GroupError::Malformed)?
            .try_into()
            .map_err(|_| GroupError::Malformed)?;
        let nonce: [u8; 12] = B64
            .decode(parts[3])
            .map_err(|_| GroupError::Malformed)?
            .try_into()
            .map_err(|_| GroupError::Malformed)?;
        let ciphertext = B64.decode(parts[4]).map_err(|_| GroupError::Malformed)?;
        Ok(Self {
            channel,
            epoch,
            ephemeral_pub,
            nonce,
            ciphertext,
        })
    }
}

/// Derive the AEAD wrapping key from an ECDH shared secret, bound to the channel
/// and epoch so a sealed blob can't be replayed onto a different channel/epoch.
fn wrap_key(shared: &[u8; 32], channel: &str, epoch: u64) -> [u8; 32] {
    let salt = Sha256::digest(channel.to_lowercase().as_bytes());
    let hk = Hkdf::<Sha256>::new(Some(&salt), shared);
    let info = format!("freeq-group-keywrap-v1-{epoch}");
    let mut key = [0u8; 32];
    hk.expand(info.as_bytes(), &mut key)
        .expect("32 is a valid HKDF-SHA256 length");
    key
}

impl GroupState {
    /// Steward helper: seal the current epoch to a set of members, returning
    /// `(member_did, EGK1-wire)` pairs ready to `POST /channels/{c}/groupkeys`
    /// as `{ epoch, keys: { member_did: wire } }`. One call per membership
    /// snapshot; call again after [`GroupState::rotate`] to re-key.
    pub fn seal_batch(&self, members: &[(String, [u8; 32])]) -> Vec<(String, String)> {
        members
            .iter()
            .map(|(did, pk)| (did.clone(), self.seal_for(pk).to_wire()))
            .collect()
    }
}

/// Member helper: from the sealed keys the server returned for us (each an
/// `(epoch, EGK1-wire)` pair, any order), recover the [`GroupState`] for the
/// **newest** epoch we can actually open. Older epochs remain available via
/// [`GroupState::open`] for decrypting history.
pub fn open_best(
    candidates: &[(u64, String)],
    my_secret: &StaticSecret,
) -> Result<GroupState, GroupError> {
    let mut sorted: Vec<&(u64, String)> = candidates.iter().collect();
    sorted.sort_by_key(|c| std::cmp::Reverse(c.0)); // newest epoch first
    for (_epoch, wire) in sorted {
        if let Ok(sealed) = SealedGroupKey::from_wire(wire)
            && let Ok(state) = GroupState::open(&sealed, my_secret)
        {
            return Ok(state);
        }
    }
    Err(GroupError::Crypto)
}

/// True if `text` is an EG1 channel message.
pub fn is_group_encrypted(text: &str) -> bool {
    text.starts_with(EG1_PREFIX)
}

/// Parse the epoch from an EG1 message without decrypting, so a member can tell
/// whether it already holds the right sealed key.
pub fn parse_epoch(wire: &str) -> Option<u64> {
    wire.strip_prefix(EG1_PREFIX)?
        .split(':')
        .next()?
        .parse()
        .ok()
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum GroupError {
    #[error("not an EG1 group message")]
    NotEg1,
    #[error("not an EGK1 sealed key")]
    NotEgk1,
    #[error("malformed message")]
    Malformed,
    #[error("invalid key")]
    BadKey,
    #[error("epoch mismatch: have {have}, got {got} (need the sealed key for that epoch)")]
    EpochMismatch { have: u64, got: u64 },
    #[error("AEAD open failed (wrong recipient, wrong key, or tampered)")]
    Crypto,
    #[error("decrypted data is not valid UTF-8")]
    Utf8,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn member() -> (StaticSecret, [u8; 32]) {
        let sk = StaticSecret::random_from_rng(OsRng);
        let pk = PublicKey::from(&sk).to_bytes();
        (sk, pk)
    }

    #[test]
    fn steward_seals_and_member_opens_then_reads() {
        let steward = GroupState::create("#eng");
        let (alice_sk, alice_pk) = member();

        // Steward seals the current epoch to Alice and posts an encrypted message.
        let sealed = steward.seal_for(&alice_pk);
        let msg = steward.encrypt("quarterly numbers attached").unwrap();

        // Server relays both blindly; Alice recovers the key and reads.
        let alice = GroupState::open(&sealed, &alice_sk).unwrap();
        assert_eq!(alice.epoch, steward.epoch);
        assert_eq!(alice.decrypt(&msg).unwrap(), "quarterly numbers attached");
    }

    #[test]
    fn secret_is_not_derivable_from_public_data() {
        // Two channels with the SAME name but independently created must NOT
        // share a key — proving the secret is random, not a function of the
        // (public) channel name the way the broken ENC2 GroupKey was.
        let a = GroupState::create("#eng");
        let b = GroupState::create("#eng");
        let wire = a.encrypt("secret").unwrap();
        // Same epoch, same channel name, but different random secret → fails.
        assert_eq!(a.epoch, b.epoch);
        assert!(matches!(b.decrypt(&wire), Err(GroupError::Crypto)));
    }

    #[test]
    fn wrong_member_cannot_open_the_seal() {
        let steward = GroupState::create("#eng");
        let (_alice_sk, alice_pk) = member();
        let (mallory_sk, _mallory_pk) = member();

        let sealed = steward.seal_for(&alice_pk);
        // Mallory was never sealed to; her static secret can't recover the key.
        assert!(matches!(
            GroupState::open(&sealed, &mallory_sk),
            Err(GroupError::Crypto)
        ));
    }

    #[test]
    fn rotation_revokes_the_departed_member() {
        // Bob is a member at epoch 1, then leaves. Steward rotates to epoch 2
        // and re-seals only to Alice. Bob cannot read epoch-2 traffic.
        let e1 = GroupState::create("#eng");
        let (alice_sk, alice_pk) = member();
        let (bob_sk, bob_pk) = member();

        let bob_e1 = GroupState::open(&e1.seal_for(&bob_pk), &bob_sk).unwrap();
        let msg_e1 = e1.encrypt("visible to bob").unwrap();
        assert_eq!(bob_e1.decrypt(&msg_e1).unwrap(), "visible to bob");

        // Bob leaves → rotate, seal only to Alice.
        let e2 = e1.rotate();
        assert_eq!(e2.epoch, 2);
        let alice_e2 = GroupState::open(&e2.seal_for(&alice_pk), &alice_sk).unwrap();
        let msg_e2 = e2.encrypt("post-offboarding secret").unwrap();

        // Alice reads epoch 2; Bob (still holding only epoch 1) cannot.
        assert_eq!(
            alice_e2.decrypt(&msg_e2).unwrap(),
            "post-offboarding secret"
        );
        assert!(matches!(
            bob_e1.decrypt(&msg_e2),
            Err(GroupError::EpochMismatch { have: 1, got: 2 })
        ));
    }

    #[test]
    fn sealed_key_wire_roundtrips() {
        let steward = GroupState::create("#secret-room");
        let (_sk, pk) = member();
        let sealed = steward.seal_for(&pk);
        let wire = sealed.to_wire();
        assert!(wire.starts_with(EGK1_PREFIX));
        assert_eq!(SealedGroupKey::from_wire(&wire).unwrap(), sealed);
    }

    #[test]
    fn tampered_ciphertext_is_rejected() {
        let steward = GroupState::create("#eng");
        let mut wire = steward.encrypt("integrity matters").unwrap();
        // Flip the last base64 char of the ciphertext.
        let last = wire.pop().unwrap();
        wire.push(if last == 'A' { 'B' } else { 'A' });
        assert!(steward.decrypt(&wire).is_err());
    }

    #[test]
    fn parse_epoch_and_detect() {
        let steward = GroupState::create("#eng").rotate().rotate(); // epoch 3
        let wire = steward.encrypt("hi").unwrap();
        assert!(is_group_encrypted(&wire));
        assert_eq!(parse_epoch(&wire), Some(3));
        assert_eq!(parse_epoch("ENC1:x:y"), None);
    }
}

File: freeq-sdk-js/src/e2ee_group.ts
/**
 * VC-bootstrapped end-to-end group encryption for channels (EG1 / EGK1).
 *
 * The web/TS counterpart of `freeq-sdk::e2ee_group` (Rust). Wire formats and
 * key derivation are byte-identical, so a Rust steward bot can seal a group key
 * that a browser member opens, and vice versa (both use RFC 7748 X25519,
 * HKDF-SHA256, and AES-256-GCM).
 *
 * Model (see docs/VC-BOOTSTRAPPED-CHANNEL-E2EE.md):
 *  - A channel has a RANDOM 32-byte group secret at a given epoch — NOT derived
 *    from any public value (that was the broken ENC2 mistake).
 *  - A steward seals the secret to each member's X25519 public key. The server
 *    stores/relays the sealed `EGK1:` blob but can never open it.
 *  - Channel traffic is `EG1:<epoch>:<nonce>:<ct>` AES-256-GCM ciphertext.
 *  - On membership change the steward rotates the epoch and re-seals to the
 *    remaining members only — the departed member cannot read new epochs.
 */

import type { ChannelCipher } from './channel-cipher.js';

const EG1_PREFIX = 'EG1:';
const EGK1_PREFIX = 'EGK1:';

// WebCrypto's lib.dom types reject `Uint8Array<ArrayBufferLike>` where a strict
// `BufferSource` is expected (SharedArrayBuffer strictness). The rest of this
// SDK works around it with an `any` view of SubtleCrypto; mirror that here.
const subtle: any = crypto.subtle;

// ── Types ──

export interface GroupState {
  channel: string;
  epoch: number;
  /** Random 32-byte group secret. Never leaves a member's device unsealed. */
  secret: Uint8Array;
}

export interface SealedGroupKey {
  channel: string;
  epoch: number;
  ephemeralPub: Uint8Array;
  nonce: Uint8Array;
  ciphertext: Uint8Array;
}

/**
 * A member's X25519 secret for opening sealed keys. WebCrypto cannot import an
 * X25519 private key from raw bytes, so pass EITHER the private `CryptoKey`
 * (the natural browser path — generate the keypair and keep it) OR the raw
 * `{ secret, publicKey }` scalar pair (for interop with keys minted elsewhere,
 * e.g. a Rust member — imported via JWK, which needs both halves).
 */
export type X25519Secret = CryptoKey | { secret: Uint8Array; publicKey: Uint8Array };

// ── Steward: group lifecycle ──

/** Create a fresh group at epoch 1 with a random secret. */
export function createGroup(channel: string): GroupState {
  return {
    channel: channel.toLowerCase(),
    epoch: 1,
    secret: crypto.getRandomValues(new Uint8Array(32)),
  };
}

/** Mint the next epoch with a new random secret (call on membership change). */
export function rotate(state: GroupState): GroupState {
  return {
    channel: state.channel,
    epoch: state.epoch + 1,
    secret: crypto.getRandomValues(new Uint8Array(32)),
  };
}

// ── Channel message encrypt / decrypt (EG1) ──

/** Encrypt a channel message → `EG1:<epoch>:<nonce>:<ct>`. */
export async function encryptGroup(state: GroupState, plaintext: string): Promise<string> {
  const key = await messageKey(state);
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const cryptoKey = await subtle.importKey('raw', key, { name: 'AES-GCM' }, false, ['encrypt']);
  const ct = new Uint8Array(await subtle.encrypt({ name: 'AES-GCM', iv }, cryptoKey, new TextEncoder().encode(plaintext)));
  return `${EG1_PREFIX}${state.epoch}:${b64(iv)}:${b64(ct)}`;
}

/** Decrypt an `EG1:` message. Returns null on wrong epoch/key/tamper. */
export async function decryptGroup(state: GroupState, wire: string): Promise<string | null> {
  if (!wire.startsWith(EG1_PREFIX)) return null;
  const parts = wire.slice(EG1_PREFIX.length).split(':');
  if (parts.length !== 3) return null;
  const epoch = Number(parts[0]);
  if (!Number.isInteger(epoch) || epoch !== state.epoch) return null;
  try {
    const iv = unb64(parts[1]);
    const ct = unb64(parts[2]);
    if (iv.length !== 12) return null;
    const key = await messageKey(state);
    const cryptoKey = await subtle.importKey('raw', key, { name: 'AES-GCM' }, false, ['decrypt']);
    const pt = await subtle.decrypt({ name: 'AES-GCM', iv }, cryptoKey, ct);
    return new TextDecoder().decode(pt);
  } catch {
    return null;
  }
}

/** True if `text` is an EG1 channel message. */
export function isGroupEncrypted(text: string): boolean {
  return text.startsWith(EG1_PREFIX);
}

/** Read the epoch off an EG1 message without decrypting. */
export function parseEpoch(wire: string): number | null {
  if (!wire.startsWith(EG1_PREFIX)) return null;
  const n = Number(wire.slice(EG1_PREFIX.length).split(':')[0]);
  return Number.isInteger(n) ? n : null;
}

// ── Key wrap: seal to a member, open with your secret (EGK1) ──

/**
 * Steward: seal this epoch's secret to a member's raw 32-byte X25519 public key.
 * Ephemeral-static ECIES — a fresh ephemeral key per seal.
 */
export async function sealFor(state: GroupState, memberPub: Uint8Array): Promise<SealedGroupKey> {
  const eph = await subtle.generateKey({ name: 'X25519' }, true, ['deriveBits']) as CryptoKeyPair;
  const ephPub = new Uint8Array(await subtle.exportKey('raw', eph.publicKey));
  const shared = await x25519(eph.privateKey, memberPub);
  const wrapKey = await deriveWrapKey(shared, state.channel, state.epoch);

  const iv = crypto.getRandomValues(new Uint8Array(12));
  const cryptoKey = await subtle.importKey('raw', wrapKey, { name: 'AES-GCM' }, false, ['encrypt']);
  const ct = new Uint8Array(await subtle.encrypt({ name: 'AES-GCM', iv }, cryptoKey, state.secret));
  return { channel: state.channel, epoch: state.epoch, ephemeralPub: ephPub, nonce: iv, ciphertext: ct };
}

/**
 * Member: recover the group state from a sealed key using your raw 32-byte
 * X25519 secret. This is the only way to obtain the secret; the server cannot.
 */
export async function openSealed(sealed: SealedGroupKey, mySecret: X25519Secret): Promise<GroupState | null> {
  try {
    const myKey = await toPrivateKey(mySecret);
    const shared = await x25519(myKey, sealed.ephemeralPub);
    const wrapKey = await deriveWrapKey(shared, sealed.channel, sealed.epoch);
    const cryptoKey = await subtle.importKey('raw', wrapKey, { name: 'AES-GCM' }, false, ['decrypt']);
    const secret = new Uint8Array(await subtle.decrypt({ name: 'AES-GCM', iv: sealed.nonce }, cryptoKey, sealed.ciphertext));
    if (secret.length !== 32) return null;
    return { channel: sealed.channel, epoch: sealed.epoch, secret };
  } catch {
    return null;
  }
}

/** Serialize to `EGK1:<channel>:<epoch>:<eph-pub>:<nonce>:<ct>`. */
export function sealedToWire(s: SealedGroupKey): string {
  return `${EGK1_PREFIX}${s.channel}:${s.epoch}:${b64(s.ephemeralPub)}:${b64(s.nonce)}:${b64(s.ciphertext)}`;
}

/** Parse an `EGK1:` control message. */
export function sealedFromWire(wire: string): SealedGroupKey | null {
  if (!wire.startsWith(EGK1_PREFIX)) return null;
  const body = wire.slice(EGK1_PREFIX.length);
  // channel : epoch : ephPub : nonce : ct  — channel has no ':' for IRC names.
  const idx: number[] = [];
  for (let i = 0, found = 0; i < body.length && found < 4; i++) {
    if (body[i] === ':') { idx.push(i); found++; }
  }
  if (idx.length !== 4) return null;
  const channel = body.slice(0, idx[0]);
  const epoch = Number(body.slice(idx[0] + 1, idx[1]));
  if (!Number.isInteger(epoch)) return null;
  try {
    return {
      channel,
      epoch,
      ephemeralPub: unb64(body.slice(idx[1] + 1, idx[2])),
      nonce: unb64(body.slice(idx[2] + 1, idx[3])),
      ciphertext: unb64(body.slice(idx[3] + 1)),
    };
  } catch {
    return null;
  }
}

// ── Convenience ──

/** Steward: seal to many members → `[member_did, EGK1-wire]` for the POST body. */
export async function sealBatch(state: GroupState, members: Array<[string, Uint8Array]>): Promise<Array<[string, string]>> {
  const out: Array<[string, string]> = [];
  for (const [did, pub] of members) out.push([did, sealedToWire(await sealFor(state, pub))]);
  return out;
}

/**
 * Member: from the sealed keys the server returned (each `[epoch, EGK1-wire]`),
 * recover the newest epoch we can open. Older epochs stay openable for history.
 */
export async function openBest(candidates: Array<[number, string]>, mySecret: X25519Secret): Promise<GroupState | null> {
  const sorted = [...candidates].sort((a, b) => b[0] - a[0]);
  for (const [, wire] of sorted) {
    const sealed = sealedFromWire(wire);
    if (!sealed) continue;
    const state = await openSealed(sealed, mySecret);
    if (state) return state;
  }
  return null;
}

// ── ChannelCipher adapter ──

/**
 * Wrap a set of group states (one per epoch we hold) as a `ChannelCipher` for
 * `FreeqClient.setChannelCipher`. `states` is a getter so the cipher always
 * sees the latest keys — install once, keep loading epochs.
 *
 * - encrypt: with the highest epoch held; `null` when none.
 * - decrypt: by the epoch named in the wire (`parseEpoch`); `null` when we
 *   do not hold it or the ciphertext does not open.
 * - isCiphertext: `isGroupEncrypted` (an `EG1:` body).
 */
export function makeGroupCipher(states: () => GroupState[]): ChannelCipher {
  const byEpoch = (): Map<number, GroupState> => {
    const m = new Map<number, GroupState>();
    for (const s of states()) m.set(s.epoch, s);
    return m;
  };
  return {
    async encrypt(plaintext: string): Promise<string | null> {
      let best: GroupState | null = null;
      for (const s of states()) if (!best || s.epoch > best.epoch) best = s;
      if (!best) return null;
      return encryptGroup(best, plaintext);
    },
    async decrypt(wire: string): Promise<string | null> {
      const epoch = parseEpoch(wire);
      if (epoch === null) return null;
      const state = byEpoch().get(epoch);
      if (!state) return null;
      return decryptGroup(state, wire);
    },
    isCiphertext(wire: string): boolean {
      return isGroupEncrypted(wire);
    },
  };
}

// ── Crypto helpers (mirror freeq-sdk-js/src/e2ee.ts) ──

async function messageKey(state: GroupState): Promise<Uint8Array> {
  const salt = new Uint8Array(await subtle.digest('SHA-256', new TextEncoder().encode(state.channel)));
  return hkdf(state.secret, salt, `freeq-group-msg-v1-${state.epoch}`);
}

async function deriveWrapKey(shared: Uint8Array, channel: string, epoch: number): Promise<Uint8Array> {
  const salt = new Uint8Array(await subtle.digest('SHA-256', new TextEncoder().encode(channel.toLowerCase())));
  return hkdf(shared, salt, `freeq-group-keywrap-v1-${epoch}`);
}

async function hkdf(ikm: Uint8Array, salt: Uint8Array, info: string): Promise<Uint8Array> {
  const base = await subtle.importKey('raw', ikm, 'HKDF', false, ['deriveBits']);
  const bits = await subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt, info: new TextEncoder().encode(info) },
    base, 256,
  );
  return new Uint8Array(bits);
}

/** Coerce a member secret into an importable X25519 private CryptoKey. */
async function toPrivateKey(sk: X25519Secret): Promise<CryptoKey> {
  if (sk instanceof CryptoKey) return sk;
  // Raw scalar pair → JWK (WebCrypto requires both d and x for OKP private keys).
  return subtle.importKey(
    'jwk',
    { kty: 'OKP', crv: 'X25519', d: b64(sk.secret), x: b64(sk.publicKey) },
    { name: 'X25519' },
    false,
    ['deriveBits'],
  );
}

async function x25519(mySecret: CryptoKey, theirPublic: Uint8Array): Promise<Uint8Array> {
  const theirKey = await subtle.importKey('raw', theirPublic, { name: 'X25519' }, false, []);
  const bits = await subtle.deriveBits({ name: 'X25519', public: theirKey }, mySecret, 256);
  return new Uint8Array(bits);
}

function b64(data: Uint8Array): string {
  return btoa(String.fromCharCode(...data)).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function unb64(str: string): Uint8Array {
  const padded = str.replace(/-/g, '+').replace(/_/g, '/') + '=='.slice(0, (4 - (str.length % 4)) % 4);
  return Uint8Array.from(atob(padded), c => c.charCodeAt(0));
}
